### PR TITLE
Split indexation into mupltiple processes

### PR DESCRIPTION
**Pull request by nikitagromov**
Pivotal story id [#143460457](https://www.pivotaltracker.com/story/show/143460457)

Also this [PR](https://github.com/geniusproject/cerri-api/pull/887) could increase performance

### CODE REVIEW

File: nefertari/scripts/es.py [hunk 1]
```diff
@@ -1,138 +1,272 @@
...
-        params = self.options.params or ''
-        params = dict([[k, v[0]] for k, v in urllib.parse.parse_qs(params).items()])
+    def __init__(self, model_names, processes, options):
+        self._pool = {process_id: [] for process_id in range(0, processes)}
+        self.options = options
 
         for model_name in model_names:
-            self.log.info('Processing model `{}`'.format(model_name))
+            for process_id in range(0, processes):
+                self._pool[process_id].append(ESTask(model_name, options))
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: Wouldn't it be simpler and more readable to reverse the `for` loop nesting and do something like this:
```python
self._pool = {}
for process_id in range(0, processes):
    self._pool[process_id] = []
    for model_name in model_names:
        self._pool[process_id].append(ESTask(model_name, options))
```

File: nefertari/scripts/es.py [hunk 2]
```diff
@@ -1,138 +1,272 @@
...
+                continue
 
-            if '_limit' not in local_params:
-                limit = model.get_collection().count()
-                local_params['_limit'] = limit
+            model_chunks[model_name] = chunks[self.id]
+        return model_chunks
 
-            chunk_size = int(self.options.chunk or local_params['_limit'])
+    def __call__(self):
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: If this is only called in `apply`, why not simply put the code in `apply`?

**Review by nmenardg — CHANGES_REQUESTED** [review]

File: nefertari/scripts/es.py (hunk 2 above)
Comment (nikitagromov) [reply nefertari/scripts/es.py]: This is wrapper, because we can't pass object method as function to `map`

File: nefertari/scripts/es.py [hunk 3]
```diff
@@ -153,107 +149,80 @@ def recreate_index(registry):
...
+            items = list(map(lambda item: getattr(item, model.pk_field()) ,sorted(query.values(model.pk_field()))))
+            chunks = list(split_collection(limit, processes_count, items))
+            model_chunks[model_name] = chunks
 
-            if len(chunks) - 1 < self.id:
-                model_chunks[model_name] = tuple()
-                continue
+            for p in range(0, processes_count):
+                if len(chunks) > p:
+                    q.put((model_name, chunks[p]), block=False)
```
Comment (nmenardg) [inline nefertari/scripts/es.py]: Do you do something with `model_chunks` afterwards? I don't quite understand what this method does

Comment (nikitagromov) [reply nefertari/scripts/es.py]: I have queue which shared between processes. This method put model chunks to queue, and other processes take model chunks from queue.

**Pull request merged**